pylops/optimization/solver.py

```diff
@@ -2,10 +2,202 @@
 
 import numpy as np
 
+from pylops.optimization.basesolver import Solver
 from pylops.utils.backend import get_array_module
 
 
-def cg(Op, y, x0, niter=10, damp=0.0, tol=1e-4, show=False, callback=None):
+class CG(Solver):
+    r"""Conjugate gradient
+
+    Solve a square system of equations given an operator ``Op`` and
+    data ``y`` using conjugate gradient iterations.
+
+    Parameters
+    ----------
+    Op : :obj:`pylops.LinearOperator`
+        Operator to invert of size :math:`[N \times N]`
+    y : :obj:`np.ndarray`
+        Data of size :math:`[N \times 1]`
+
+    Attributes
+    ----------
+    x : :obj:`np.ndarray`
+        Estimated model of size :math:`[N \times 1]`
+    iit : :obj:`int`
+        Number of executed iterations
+    cost : :obj:`numpy.ndarray`, optional
+        History of the L2 norm of the residual
+
+    Notes
+    -----
+    Solve the :math:`\mathbf{y} = \mathbf{Opx}` problem using conjugate gradient
+    iterations [1]_.
+
+    .. [1] Hestenes, M R., Stiefel, E., “Methods of Conjugate Gradients for Solving
+       Linear Systems”, Journal of Research of the National Bureau of Standards.
+       vol. 49. 1952.
+
+    """
+
+    def __init__(self, Op, y):
+        self.Op = Op
+        self.y = y
+        self.ncp = get_array_module(y)
+
+    def _print_setup(self):

```

Comment:
mmh apart from `_print_finalize (which I moved to the base class) I am not so convinced about the benefit of doing this... there is quite some difference between solvers. 

Ok, they all instantiate start and print some similar lines but eventually they diverge quite a lot. I worry we won't gain much. Let's get all solvers in place and then we can look at

Reply:
Okay but I'm referring to the *formatting*, not the content. Like the linewidth, how many decimal places and spacings between columns and stuff like

```
print("-----------------------------------------------------------")
```

Reply:
Then I don’t understand… you mean we could have

print(‘-‘*N) 

but then what do we do we make a _print_line() in the base class and call it when we want a line… is this what you have in mind?

Reply:
I guess it is hard to say when only a single solver is in the PR. Just leave it out from this PR and when all Solvers are class based it can be refactored. Just an idea on my end, to separate the formatting from the content. Essentially, the specific solver should never call print() but be handled in the base class. I could probably have a go at it later.

Reply:
Alright, I am going to change all solvers once we agree on the format. I won't merge this PR until that is done. 

So for now I will keep this comment unsolved and we can pick it up again and see what can be done

Reply:
Sorry for being late to the party, but I think TensorFlow is a good inspiration for this.  Their `model.fit` method takes a verbose option, which is pretty much the same across all solvers. In their case they show things like iterations/epochs. 

In our case, solvers are completely free to process the data in any way they like. In this case, I don't see a benefit in having a base class method handling it. For more complex printing, TensorFlow uses callbacks which can be register for any kind of event, including printing (methods called "on_step", "on_solve", "on_setup", etc.). I think we could probably benefit from this approach here.

Reply:
Alright, I do not have much experience with Tensorflow but I had a look at it (https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit) and I think I get your point. Let me try to see if we agree:

- their `model.fit` is going to be a bit like our `solve`. I think this is Keras (ok now it is all in Tensorflow) so I compare it to our `solve` method... we actually allow now two levels of 'complexity': high level, Keras like, or low level with `step` and letting user do the for..loop and anything else, Torch or pure Tensorflow like. Makes sense?
- so within `solve` we have a `show` (we always had it and I would not change it into verbose to be consistent with spicy which is the API we are the most aligned with) and now`itershow=[N1, N2, N3]` which allows choosing at which iterations to show the progress (e.g. first N1, last N2, and every N3 in between). **Do you feel we should have more degrees of verbosity/showing like in Tensorflow (even though they also say people should use 0 or 2..)?**
- for the callbacks, we currently allow only one, are you suggesting we should allow a list? In tensorflow, it looks to me you have a Callback class and you can register multiple callbacks. The full list is here, https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#expandable-1 (I don't see the ones you mention unless I am looking in the wrong place) and then inside fit those are called at specific places... are you suggesting we do something along these lines? This sounds interesting but I find it a bit different from the  _print_xxx() we are discussing here, unless you think **we should also turn also all prints into callbacks?** We could then have predefined callbacklists for all solvers, and allow user to instantiate them and modify any of the callbacks. To me we may be up to making something more complicated that it should be as we provide pretty much custom printing for every solver and I don't see why someone would want to change them (but maybe I am wrong ;) ). So we currently have:

     * _print_setup
     * _print_step
     * _print_finalize
     * callback

**Do you see the need for more, any suggestion?**

Reply:
* Yup, I think it is a good approach.
* I'm OK with the current verbosity levels, but this can be changed by callbacks, which lead me to:
* Yes, TensorFlow implements a callback class, which could be overhead for us, but they get an extremely flexible "engine" with it. Every TF callback has methods like "on_train_end", "on_epoch_begin" and [things like that](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#methods). This gives the user fine-gained introspection in case they want to print/store/manipulate any part of the fitting procedure. Where I think we can benefit from this: I was envisioning that each solver can have an "on_(setup|step|solve)_(begin|end)" list of callbacks. Now, the reason these callbacks should probably be a class and not a function is because with a class you can store information from any and all steps/iterations/solves.
   Here is an example: I would like to lsqr solve and print the percentage change in the norm of the data residual. This requires that I keep information from the previous steps. Another example: I want a "patience"-like stopper related to this gradient norm: if the data residual is increasing at a slow pace, stop or change something. The nice thing about this is that verbosity can be changed very easily by creating a LoggerCB(verb=0,1,2,3,...) that just implements prints.
   I know that many of the callbacks they implement are related to the nonlinear nature of the solvers, but not all of them. If we think we can benefit from some of them (see [some default ones](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/)), I think it would be worth the effort (and overhead). None of this would change the fact that the simple function interface should still be available (and be very easy to use!).